File: skills/math-expression-parser/reference/src/ast-types.ts
/**
 * AST node type definitions for parsed math expressions.
 *
 * The AST is a tree of three node types:
 * - NumberLiteral: a numeric value (leaf)
 * - UnaryExpr: a unary operator applied to an operand (e.g., -x)
 * - BinaryExpr: a binary operator applied to left and right operands
 *
 * @node ast-types
 * @contract ast-types.test.ts
 * @hint types: These are plain data types with no behavior. Use tagged
 *       unions, sealed classes, or enum variants in the target language.
 */

export type BinaryOp = "+" | "-" | "*" | "/" | "%" | "**";
export type UnaryOp = "-";

export interface NumberLiteral {
  type: "number";
  value: number;
}

export interface UnaryExpr {
  type: "unary";
  op: UnaryOp;
  operand: AstNode;
}

export interface BinaryExpr {
  type: "binary";
  op: BinaryOp;
  left: AstNode;
  right: AstNode;
}

export type AstNode = NumberLiteral | UnaryExpr | BinaryExpr;

export function numberLiteral(value: number): NumberLiteral {
  return { type: "number", value };
}

export function unaryExpr(op: UnaryOp, operand: AstNode): UnaryExpr {
  return { type: "unary", op, operand };
}

export function binaryExpr(
  op: BinaryOp,
  left: AstNode,
  right: AstNode,
): BinaryExpr {
  return { type: "binary", op, left, right };
}

File: research/experiments/sweep/runs/mathexpr-full-kotlin-sonnet45/src/main/kotlin/mathexpr/AstTypes.kt
package mathexpr

sealed class AstNode {
    data class NumberLiteral(val value: Double) : AstNode()
    data class UnaryExpr(val op: String, val operand: AstNode) : AstNode()
    data class BinaryExpr(val op: String, val left: AstNode, val right: AstNode) : AstNode()
}

fun numberLiteral(value: Double): AstNode.NumberLiteral {
    return AstNode.NumberLiteral(value)
}

fun unaryExpr(op: String, operand: AstNode): AstNode.UnaryExpr {
    return AstNode.UnaryExpr(op, operand)
}

fun binaryExpr(op: String, left: AstNode, right: AstNode): AstNode.BinaryExpr {
    return AstNode.BinaryExpr(op, left, right)
}
